bson/decoder.go

```diff
@@ -79,92 +73,15 @@ func (d *Decoder) Decode(val interface{}) error {
 	}
-	decoder, err := d.dc.LookupDecoder(rval.Type())
+	decoder, err := d.reg.LookupDecoder(rval.Type())
 	if err != nil {
 		return err
 	}
 
-	if d.defaultDocumentM {
-		d.dc.DefaultDocumentM()
-	}
-	if d.defaultDocumentD {
-		d.dc.DefaultDocumentD()
-	}
-	if d.binaryAsSlice {
-		d.dc.BinaryAsSlice()
-	}
-	if d.useJSONStructTags {
-		d.dc.UseJSONStructTags()
-	}
-	if d.useLocalTimeZone {
-		d.dc.UseLocalTimeZone()
-	}
-	if d.zeroMaps {
-		d.dc.ZeroMaps()
-	}
-	if d.zeroStructs {
-		d.dc.ZeroStructs()
-	}
-
-	return decoder.DecodeValue(d.dc, d.vr, rval)
-}
-
-// Reset will reset the state of the decoder, using the same *DecodeContext used in
-// the original construction but using vr for reading.
-func (d *Decoder) Reset(vr ValueReader) {
-	d.vr = vr
-}
-
-// SetRegistry replaces the current registry of the decoder with r.
-func (d *Decoder) SetRegistry(r *Registry) {
-	d.dc.Registry = r
-}
-
-// DefaultDocumentM causes the Decoder to always unmarshal documents into the primitive.M type. This
-// behavior is restricted to data typed as "interface{}" or "map[string]interface{}".
-func (d *Decoder) DefaultDocumentM() {
-	d.defaultDocumentM = true
-}
-
-// DefaultDocumentD causes the Decoder to always unmarshal documents into the primitive.D type. This
-// behavior is restricted to data typed as "interface{}" or "map[string]interface{}".
-func (d *Decoder) DefaultDocumentD() {
-	d.defaultDocumentD = true
-}
-
-// AllowTruncatingDoubles causes the Decoder to truncate the fractional part of BSON "double" values
-// when attempting to unmarshal them into a Go integer (int, int8, int16, int32, or int64) struct
-// field. The truncation logic does not apply to BSON "decimal128" values.
-func (d *Decoder) AllowTruncatingDoubles() {
-	d.dc.Truncate = true
-}
-
-// BinaryAsSlice causes the Decoder to unmarshal BSON binary field values that are the "Generic" or
-// "Old" BSON binary subtype as a Go byte slice instead of a primitive.Binary.
-func (d *Decoder) BinaryAsSlice() {
-	d.binaryAsSlice = true
-}
-
-// UseJSONStructTags causes the Decoder to fall back to using the "json" struct tag if a "bson"
-// struct tag is not specified.
-func (d *Decoder) UseJSONStructTags() {
-	d.useJSONStructTags = true
-}
-
-// UseLocalTimeZone causes the Decoder to unmarshal time.Time values in the local timezone instead
-// of the UTC timezone.
-func (d *Decoder) UseLocalTimeZone() {
-	d.useLocalTimeZone = true
-}
-
-// ZeroMaps causes the Decoder to delete any existing values from Go maps in the destination value
-// passed to Decode before unmarshaling BSON documents into them.
-func (d *Decoder) ZeroMaps() {
-	d.zeroMaps = true
+	return decoder.DecodeValue(d.reg, d.vr, rval)
 }
 
-// ZeroStructs causes the Decoder to delete any existing values from Go structs in the destination
-// value passed to Decode before unmarshaling BSON documents into them.
-func (d *Decoder) ZeroStructs() {
-	d.zeroStructs = true
```

Comment:
Removing these configuration methods and replacing them with `SetBehavior` is a major refactor of the `bson.Decoder` API and seems outside the scope of [GODRIVER-2714](https://jira.mongodb.org/browse/GODRIVER-2714). What motivated the change to this API?